### PR TITLE
Drop the `Property` trait entirely

### PR DESCRIPTION
**Pull request by apoelstra**
Something of a followup to #649, which specifically removes the `type_check` method from `Property`.

Basically, this whole trait is confused. Originally it was supposed to express "a recursively defined property of a Miniscript" which could be defined by a couple constructors and then automatically computed for any Miniscript. In fact, the trait required a separate constructor for every single Miniscript fragment so nothing was "automatic" except sometimes reusing code between the different hash fragments and between `older`/`after`.

Furthermore, every implementor of this trait had slightly different requirements, meaning that there were `unreachable!()`s throughout the code, functions that were never called (e.g. `Type::from_sha256` passing through to `Correctness::from_sha256` and `Malleability::from_sha256` when all three of those types implemented the function by calling `from_hash`, which *also* was defined by `Type::from_hash` calling `Correctness::from_hash` and `Malleability::from_hash`. So much boilerplate) and many instances of methods being implemented but ignoring arguments, or (worse) being generic over `Pk`/`Ctx` and ignoring those types entirely, or returning `Result` but always returing the `Ok` variant.

So basically, there was no value in being generic over the trait and the trait wasn't even a generalization of any of the types that implemented it.

**Adding to this** having the trait prevents us from having constfns in our type checking code, which sucks because we have a lot of common fixed fragments. It also prevents us from returning different error types from different parts of the typechecker, which is my specific goal in removing the trait.

This PR has a lot of commits but most of them are mechanical and/or small.

### CODE REVIEW

**Review by sanket1729 — APPROVED** [review]
ACK e971e77d64527d8f70bad690500ea12a51cfdf08. Looks a lot cleaner. Thanks for these changes. It's nice that we no longer have the artificial error paths because the parent trait had the definition to encompass all possible implementations.

File: src/miniscript/types/correctness.rs [hunk 1]
```diff
@@ -95,16 +106,15 @@ impl Correctness {
...
             && self.input.is_subtype(other.input)
             && self.dissatisfiable >= other.dissatisfiable
             && self.unit >= other.unit
     }
-}
 
-impl Property for Correctness {
-    fn sanity_checks(&self) {
+    /// Confirm invariants of the correctness checker.
+    pub fn sanity_checks(&self) {
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply src/miniscript/types/correctness.rs:117]: Sounds good! I'll do this in a followup.

[the remark this answers is not in the document]
Comment (apoelstra) [reply src/miniscript/types/correctness.rs:117]: Actually I'll just file an issue. There are a lot of `sanity_check` methods in the code, many of which are not called from within the library, and at least one has a doc example suggesting to use it to check "whether all spend paths are accessible in the Bitcoin network".

So we need to re-assess all these functions and see if they can be pulled into the type system somehow so you simply can't create the objects without running the (non-pub) methods.
